src/compiler.ts

```diff
@@ -4617,79 +4619,225 @@ export class Compiler extends DiagnosticEmitter {
+              if (this.options.isWasm64) {
+                this.i64PowInstance = instance = this.resolver.resolveFunction(<FunctionPrototype>prototype, null);
+              } else {
+                this.i32PowInstance = instance = this.resolver.resolveFunction(<FunctionPrototype>prototype, null);
+              }
+            }
+            if (!instance || !this.compileFunction(instance)) {
               expr = module.unreachable();
-              break;
+            } else {
+              expr = this.makeCallDirect(instance, [ leftExpr, rightExpr ], expression);
             }
-            let prototype = namespace.members ? namespace.members.get(CommonNames.pow) : null;
-            if (!prototype) {
-              this.error(
-                DiagnosticCode.Cannot_find_name_0,
-                expression.range, "Math.pow"
+            break;
+          }
+          */
+          case TypeKind.F32: {
+            instance = this.f32PowInstance;
+            if (!instance) {
+              let namespace = this.program.lookupGlobal(CommonNames.Mathf);
+              if (!namespace) {
+                this.error(
+                  DiagnosticCode.Cannot_find_name_0,
+                  expression.range, "Mathf"
+                );
+                expr = module.unreachable();
+                break;
+              }
+              let prototype = namespace.members ? namespace.members.get(CommonNames.pow) : null;
+              if (!prototype) {
+                this.error(
+                  DiagnosticCode.Cannot_find_name_0,
+                  expression.range, "Mathf.pow"
+                );
+                expr = module.unreachable();
+                break;
+              }
+              assert(prototype.kind == ElementKind.FUNCTION_PROTOTYPE);
+              this.f32PowInstance = instance = this.resolver.resolveFunction(<FunctionPrototype>prototype, null);
+            }
+            if (!instance || !this.compileFunction(instance)) {
+              expr = module.unreachable();
+            } else {
+              expr = this.makeCallDirect(instance, [ leftExpr, rightExpr ], expression);
+            }
+            break;
+          }
+          // Math.pow otherwise (result is f64)
+          // TODO: should the result be converted back?
+
+          // FIXME: tmp workaround
+          case TypeKind.ISIZE:
+          case TypeKind.USIZE:
+          case TypeKind.F64: {
+            // FIXME: tmp workaround
+            if (this.currentType.kind == TypeKind.ISIZE || this.currentType.kind == TypeKind.USIZE) {
+              leftExpr = this.convertExpression(leftExpr,
+                this.currentType, Type.f64,
+                false, false,
+                left
               );
+              leftType = this.currentType;
+              rightExpr = this.compileExpression(right, Type.f64, Constraints.CONV_IMPLICIT);
+              rightType = this.currentType;
+            }
```

Comment:
@dcodeIO This workaround exists due to reason that test (2 ** 2).toString() == "4.0" in `resolve-binary` trigger `F64#toString()` instead `I32#toString()`. Could you check this?

Reply:
Perhaps the resolver needs to be updated as well, so when`2 ** 2` is resolved, the correct type is returned by it. Would imagine that it currently still returns `f64`, hence picking the wrong `toString` implementation.